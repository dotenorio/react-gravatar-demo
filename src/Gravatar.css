form label {
  display: inline-block;
}
form input {
  margin-left: 5px;
  border: 1px solid #999;
  border-radius: 3px;
  display: inline-block;
  padding: 5px;
  font-size: 16px;
}
form input[type=submit] {
  border: none;
  background-color: #37a15f;
  color: #FFF;
  padding: 7px;
  border-radius: 3px;
  font-size: 14px;
  display: inline-block;
}

.result {
  margin-top: 40px;
}

pre {
  background-color: #CCC;
  padding: 15px;
  border-radius: 3px;
  box-shadow: 1px 1px 5px #000;
  white-space: pre-wrap;
  word-break: break-word;
}

.thumbnailUrl {
  position: absolute;
  right: 25px;
  top: 105px;
  border: 1px solid #CCC;
  border-radius: 5px;
}

@media only screen and (max-width: 375px) {
  form input[type=text] {
    width: 150px;
  }
}

@media only screen and (max-width: 768px) {
  form {
    text-align: center;
  }
  .thumbnailUrl {
    position: initial;
    right: auto;
    top: auto;
    border: 1px solid #CCC;
    border-radius: 5px;
    display: block;
    margin: 0 auto;
    margin-bottom: 40px;
  }
}