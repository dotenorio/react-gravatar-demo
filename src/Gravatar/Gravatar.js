import React, { Component } from 'react';
import md5 from 'md5';
import jsonp from 'universal-jsonp';

import Layout from "../Layout/Layout";
import "./Gravatar.css";

function Result(props) {
  return (
    <div className="result">
      {props.res.thumbnailUrl &&
        <img src={props.res.thumbnailUrl} alt={props.res.preferredUsername} className="thumbnailUrl" />
      }
      <pre>
        {JSON.stringify(props.res, null, 2)}
      </pre>
    </div>
  );
}

class EmailForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      res: {
        HowTo: 'Insert your email above to see results!'
      }
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      value: event.target.value
    });
  }

  handleRes(res) {
    this.setState({ res });    
  }

  async handleSubmit(event) {
    event.preventDefault();
    const valueMd5 = md5(this.state.value);
    this.handleRes({
      res: {
        status: 'Loading..'
      }
    });
    try {
      const request = await jsonp(`https://www.gravatar.com/${valueMd5}.json`, { timeout: 10000 })
      const response = await request.json();
      this.handleRes(response.entry[0]);
    } catch (err) {
      this.handleRes(err.message);
    }
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <label>
            Email:
            <input type="text" value={this.state.value} onChange={this.handleChange} placeholder="Your e-mail here.." />
          </label>
          <input type="submit" value="Enviar" />
        </form>
        <Result res={this.state.res} />
      </div>
    );
  }
}

function Gravatar () {
  return (
    <div>
      <Layout>
        <EmailForm />
      </Layout>
    </div>
  );
}

export default Gravatar;
