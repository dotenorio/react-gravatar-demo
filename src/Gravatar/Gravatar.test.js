import React from 'react';
import { MemoryRouter } from 'react-router-dom'
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import jsonp from 'universal-jsonp';
import Gravatar from './Gravatar';

jest.mock('universal-jsonp')

test('Load Gravatar', async () => {  
  const wrapper = shallow(
    <Gravatar />
  );

  const entry = [
    {
      "preferredUsername": "dotenorio",
      "thumbnailUrl": "https://secure.gravatar.com/avatar/9c62a4730ba4ebf764862b7866038dcb"
    }
  ]

  jsonp.mockImplementationOnce(() => Promise.resolve({
    json: () => Promise.resolve({
      entry
    }),
    ok: true
  }));

  jsonp.mockImplementationOnce(() => Promise.resolve({
    ok: false
  }));
  
  const EmailForm = wrapper.find('EmailForm');
  const EmailFormWraper = EmailForm.shallow();
  const Result = EmailFormWraper.find('Result');
  const ResultWrapper = Result.shallow();
  
  expect(EmailForm.length).toEqual(1);
  expect(EmailFormWraper.find('label').length).toEqual(1);
  expect(EmailFormWraper.find('label').text()).toEqual('Email:');
  expect(EmailFormWraper.find('input[type="text"]').length).toEqual(1);
  expect(EmailFormWraper.find('input[type="submit"]').length).toEqual(1);

  expect(Result.length).toEqual(1);
  expect(ResultWrapper.find('pre').text()).toContain('Insert your email above to see results!');
  
  // ToDo ~> Quebrar os blocos abaixo em 3 testes diferentes
  
  // Success
  EmailFormWraper.find('input[type="text"]').simulate('change', { target: { value: 'dotenorio@gmail.com'} });
  EmailFormWraper.find('form').simulate('submit', { preventDefault: () => { } });
  
  // Success :: With Image
  ResultWrapper.setProps({ res: entry[0] });
  expect(ResultWrapper.find('img').length).toEqual(1);

  // Forced Error
  EmailFormWraper.find('input[type="text"]').simulate('change', { target: { value: 'invalid@mail.com'} });
  EmailFormWraper.find('form').simulate('submit', { preventDefault: () => { } });
})

test('Gravatar renders correctly', () => {
  const tree = renderer
    .create(<MemoryRouter><Gravatar /></MemoryRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});