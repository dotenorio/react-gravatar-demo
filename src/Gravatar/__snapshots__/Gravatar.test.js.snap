// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gravatar renders correctly 1`] = `
<div>
  <div>
    <div
      className="header"
    >
      <h1>
        React Gravatar Demo
      </h1>
      <ul>
        <li>
          <a
            aria-current="page"
            className="active"
            href="/"
            onClick={[Function]}
            style={Object {}}
            title="Home Page"
          >
            Home
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            href="/about"
            onClick={[Function]}
            title="About Project"
          >
            About
          </a>
        </li>
      </ul>
    </div>
    <div
      className="content"
    >
      <div>
        <form
          onSubmit={[Function]}
        >
          <label>
            Email:
            <input
              onChange={[Function]}
              placeholder="Your e-mail here.."
              type="text"
              value=""
            />
          </label>
          <input
            type="submit"
            value="Enviar"
          />
        </form>
        <div
          className="result"
        >
          <pre>
            {
  "HowTo": "Insert your email above to see results!"
}
          </pre>
        </div>
      </div>
    </div>
  </div>
</div>
`;
